// vscode-markdown-languageservice/out/workspaceCache.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
import { CancellationTokenSource } from 'vscode-languageserver-protocol';
import { getDocUri } from './types/textDocument.js';
import { Disposable } from './util/dispose.js';
import { lazy } from './util/lazy.js';
import { ResourceMap } from './util/resourceMap.js';
/**
 * Cache of information per-document in the workspace.
 *
 * The values are computed lazily and invalidated when the document changes.
 */
export class MdDocumentInfoCache extends Disposable {
    #cache = new ResourceMap();
    #loadingDocuments = new ResourceMap();
    #workspace;
    #getValue;
    constructor(workspace, getValue) {
        super();
        this.#workspace = workspace;
        this.#getValue = getValue;
        this._register(this.#workspace.onDidChangeMarkdownDocument(doc => this.#invalidate(doc)));
        this._register(this.#workspace.onDidDeleteMarkdownDocument(this.#onDidDeleteDocument, this));
    }
    async get(resource) {
        let existing = this.#cache.get(resource);
        if (existing) {
            return existing.value.value;
        }
        const doc = await this.#loadDocument(resource);
        if (!doc) {
            return undefined;
        }
        // Check if we have invalidated
        existing = this.#cache.get(resource);
        if (existing) {
            return existing.value.value;
        }
        return this.#resetEntry(doc)?.value;
    }
    getForDocument(document) {
        const existing = this.#cache.get(getDocUri(document));
        if (existing) {
            return existing.value.value;
        }
        return this.#resetEntry(document).value;
    }
    #loadDocument(resource) {
        const existing = this.#loadingDocuments.get(resource);
        if (existing) {
            return existing;
        }
        const p = this.#workspace.openMarkdownDocument(resource);
        this.#loadingDocuments.set(resource, p);
        p.finally(() => {
            this.#loadingDocuments.delete(resource);
        });
        return p;
    }
    #resetEntry(document) {
        const uri = getDocUri(document);
        const existing = this.#cache.get(uri);
        if (existing) {
            existing.cts.cancel();
            existing.cts.dispose();
        }
        const cts = new CancellationTokenSource();
        const value = lazy(() => this.#getValue(document, cts.token));
        this.#cache.set(uri, { value, cts });
        return value;
    }
    #invalidate(document) {
        if (this.#cache.has(getDocUri(document))) {
            this.#resetEntry(document);
        }
    }
    #onDidDeleteDocument(resource) {
        this.#evict(resource);
    }
    invalidate(resource) {
        this.#evict(resource);
    }
    #evict(resource) {
        const entry = this.#cache.get(resource);
        if (entry) {
            entry.cts.cancel();
            entry.cts.dispose();
            this.#cache.delete(resource);
        }
    }
}
/**
 * Cache of information across all markdown files in the workspace.
 *
 * Unlike {@link MdDocumentInfoCache}, the entries here are computed eagerly for every file in the workspace.
 * However the computation of the values is still lazy.
 */
export class MdWorkspaceInfoCache extends Disposable {
    #cache = new ResourceMap();
    #init;
    #workspace;
    #getValue;
    constructor(workspace, getValue) {
        super();
        this.#workspace = workspace;
        this.#getValue = getValue;
        this._register(this.#workspace.onDidChangeMarkdownDocument(this.#onDidChangeDocument, this));
        this._register(this.#workspace.onDidCreateMarkdownDocument(this.#onDidChangeDocument, this));
        this._register(this.#workspace.onDidDeleteMarkdownDocument(this.#onDidDeleteDocument, this));
    }
    async entries() {
        await this.#ensureInit();
        return Promise.all(Array.from(this.#cache.entries(), async ([k, v]) => {
            return [k, await v.value.value];
        }));
    }
    async values() {
        await this.#ensureInit();
        return Promise.all(Array.from(this.#cache.entries(), x => x[1].value.value));
    }
    async getForDocs(docs) {
        for (const doc of docs) {
            if (!this.#cache.has(getDocUri(doc))) {
                this.#update(doc);
            }
        }
        return Promise.all(docs.map(doc => this.#cache.get(getDocUri(doc)).value.value));
    }
    async #ensureInit() {
        if (!this.#init) {
            this.#init = this.#populateCache();
        }
        await this.#init;
    }
    async #populateCache() {
        const markdownDocuments = await this.#workspace.getAllMarkdownDocuments();
        for (const document of markdownDocuments) {
            if (!this.#cache.has(getDocUri(document))) {
                this.#update(document);
            }
        }
    }
    #update(document) {
        // TODO: cancel old request?
        const cts = new CancellationTokenSource();
        this.#cache.set(getDocUri(document), {
            value: lazy(() => this.#getValue(document, cts.token)),
            cts
        });
    }
    #onDidChangeDocument(document) {
        this.#update(document);
    }
    #onDidDeleteDocument(resource) {
        const entry = this.#cache.get(resource);
        if (entry) {
            entry.cts.cancel();
            entry.cts.dispose();
            this.#cache.delete(resource);
        }
    }
}
